### PR TITLE
Streaming ingest

### PR DESCRIPTION
**Pull request by AsafMah**
-Upgrade go version to 1.16 (1.13 is OOL for a while, and we need new features)
-Adding streaming ingest client with the same ability as ingest.
-Deprecating the Stream method inside ingest
-"Modernize" e2e testing - 
* Replaced "github.com/kylelemons/godebug/pretty" with `testify` because:
    * pretty is subtly wrong - it will show an object and a pointer to the object as equal
    * The output of testify is easier to read, for both humans and for IDEs (goland for example can understand the diff)
    * pretty is pretty much dead
* Making the tests run with Subtests (now you can know which test in the table failed easily)
* Later we can go over the rest of the code and modernize other tests

Still needed:

- Docs and samples

### CODE REVIEW

File: kusto/ingest/internal/conn/conn.go [hunk 1]
```diff
@@ -122,8 +124,13 @@ func (c *Conn) Write(ctx context.Context, db, table string, payload *bytes.Buffe
 		c.headersPool <- copyHeaders(c.reqHeaders)
 	}()
 
+	if clientRequestId != "" {
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/ingest/internal/conn/conn.go:127]: Strings in go are never nil

File: kusto/ingest/ingest.go [hunk 2]
```diff
@@ -464,22 +197,20 @@ func (i *Ingestion) Stream(ctx context.Context, payload []byte, format DataForma
...
+		},
 	}
 
-	if err := zw.Close(); err != nil {
-		return errors.E(errors.OpIngestStream, errors.KClientArgs, err).SetNoRetry()
-	}
-	if buf.Len() > 4*mib {
-		return ErrTooLarge
-	}
+	_, err = streamImpl(i.db, i.table, c, ctx, bytes.NewReader(payload), props)
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/ingest/ingest.go:211]: yes

File: kusto/ingest/file_options.go [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+	return o.name
+}
+
+func (o option) OptionScopes() OptionScope {
+	return o.scopes
+}
+
+func (o option) Run(p *properties.All, scopes OptionScope) error {
+
+	for scope, scopeName := range scopesMap {
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/ingest/file_options.go]: Done
Let me know if that what you had in mind

File: kusto/ingest/file_options.go [hunk 4]
```diff
@@ -9,57 +9,58 @@ import (
...
-	QueuedIngest:     "QueuedIngest",
-	StreamingIngest:  "StreamingIngest",
-}
-
 // FileOption is an optional argument to FromFile().
 type FileOption interface {
 	fmt.Stringer
 
-	OptionScopes() OptionScope
+	TypeScopes() TypeScope
```
Comment (yogilad) [inline kusto/ingest/file_options.go:27]: SourceScope since our terminology is SourceOptions etc

File: kusto/ingest/file_options.go [hunk 5]
```diff
@@ -9,57 +9,58 @@ import (
...
-}
+func (o option) Run(p *properties.All, clientType ClientScope, ingestionType TypeScope) error {
+	errType := errors.OpFileIngest
+	if clientType&StreamingClient != 0 {
+		errType = errors.OpIngestStream
+	}
 
-func (o option) Run(p *properties.All, scopes OptionScope) error {
+	if o.clientScopes&clientType == 0 {
+		return errors.ES(errType, errors.KClientArgs, fmt.Sprintf("%s is not valid for client type", o.name))
```
Comment (yogilad) [inline kusto/ingest/file_options.go:59]: Can you print the current client type and ingestion type (need some to string API)

**Review by yogilad — APPROVED** [review]

File: kusto/ingest/file_options.go (hunk 4 above)
Comment (AsafMah) [reply kusto/ingest/file_options.go]: Done

File: kusto/ingest/file_options.go (hunk 5 above)
Comment (AsafMah) [reply kusto/ingest/file_options.go]: Done

**Pull request merged**